### PR TITLE
Jax support

### PR DESCRIPTION
**Pull request by Mivg**
Added JAX support for DoG

### CODE REVIEW

File: src/dog/jax/dog.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+@jax.jit
+def _sum_params(params: base.Params) -> chex.Array:
+    """Compute sum(params)."""
+    sums = jax.tree_util.tree_map(jnp.sum, params)
+    return jax.tree_util.tree_reduce(operator.add, sums)
+
+
+@jax.jit
+def _params_squared_norm(params: base.Params, other: Optional[base.Params]=None) -> chex.Array:
```
Comment (aviclu) [inline src/dog/jax/dog.py:20]: Have you considered using jax.numpy.linalg.norm instead?

File: example_jax.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        metrics.append(loss)
+        if i % log_interval == 0:
+            print(f'Epoch: {epoch}, Step: {i}, Loss: {np.mean(metrics)}')
+            metrics = []
+    return state, {'loss': np.mean(metrics)}
+
+
+def compute_loss(logits, labels):
+    """Compute the loss."""
+    return -jnp.mean(jnp.sum(labels * logits, axis=-1))
```
Comment (aviclu) [inline example_jax.py]: Would look better to wrap with jax.lax.neg instead of minus.
Also, you might want to move to (labels * logits).sum(-1).mean(), more common

File: example_jax.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+
+def compute_accuracy(logits, labels):
+    """Compute the accuracy."""
+    return jnp.mean(jnp.argmax(logits, -1) == jnp.argmax(labels, -1))
```
Comment (aviclu) [inline example_jax.py]: jnp.mean(jnp.equal(jnp.argmax(logits, axis=-1), targets))

File: src/dog/jax/dog.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    """Compute the l2 norm ||params - other||^2 if other is given, otherwise compute ||params||^2."""
+    if other is not None:
+        params = jax.tree_util.tree_map(lambda x, y: x - y, params, other)
+    return _sum_params(jax.tree_map(lambda x: jnp.sum(x ** 2), params))
+
+
+@jax.jit
+def _params_norm(params: base.Params, other: Optional[base.Params]=None) -> chex.Array:
+    """Compute the l2 norm ||params - other|| if other is given, otherwise compute ||params||."""
+    return jnp.sqrt(_params_squared_norm(params, other))
```
Comment (aviclu) [inline src/dog/jax/dog.py:31]: Why not tree_reduce in here?

File: src/dog/jax/dog.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    def update_fn(updates, state, params=None):
+        # assuming updates are the gradients (and that they were not scaled yet)
+        if weight_decay > 0.0:
+            raise NotImplementedError('weight decay is not implemented yet')
+            # updates = jax.tree_multimap(lambda p, g: g + weight_decay * p, params, updates)
+
+        def first_update(_):
+            init_buffer = jax.tree_util.tree_map(jnp.copy, params)
+            params_norm = _params_norm(params)
+            rbar = reps_rel * (1 + params_norm)
```
Comment (aviclu) [inline src/dog/jax/dog.py:74]: tree map and reduce?

File: src/dog/jax/dog.py (hunk 1 above)
Comment (Mivg) [reply src/dog/jax/dog.py:20]: Done, thanks
